x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/HttpCertificateCommand.java

```diff
@@ -0,0 +1,1149 @@
+            terminal.println("manual update a few months before it expires.");
+
+            validity = readPeriodInput(terminal, "CA Validity", validity);
+
+            printHeader("What key size should your CA have?", terminal);
+            terminal.println("The RSA private key for your Certificate Authority has a fixed 'key size' (in bits).");
+            terminal.println("Larger key sizes are generally more secure, but are also slower.");
+            terminal.println("");
+            terminal.println("We recommend that you use one of 2048, 3072 or 4096 bits for your key.");
+
+            keySize = readKeySize(terminal, keySize);
+            terminal.println("");
+        }
+
+        try {
+            final KeyPair keyPair = CertGenUtils.generateKeyPair(keySize);
+            final ZonedDateTime notBefore = ZonedDateTime.now(ZoneOffset.UTC);
+            final ZonedDateTime notAfter = notBefore.plus(validity);
+            X509Certificate caCert = generateSignedCertificate(dn, null, keyPair, null, null, true, notBefore, notAfter, null);
+
+            printHeader("CA password", terminal);
+            terminal.println("We recommend that you protect your CA private key with a strong password.");
+            terminal.println("If your key does not have a password (or the password can be easily guessed)");
+            terminal.println("then anyone who gets a copy of the key file will be able to generate new certificates");
+            terminal.println("and impersonate your Elasticsearch cluster.");
+            terminal.println("");
+            terminal.println("IT IS IMPORTANT THAT YOU REMEMBER THIS PASSWORD AND KEEP IT SECURE");
+            terminal.println("");
+            final char[] password = readPassword(terminal, "CA password: ", true);
+            return new CertificateTool.CAInfo(caCert, keyPair.getPrivate(), true, password.length == 0 ? null : password);
+        } catch (GeneralSecurityException | CertIOException | OperatorCreationException e) {
+            throw new IllegalArgumentException("Cannot generate CA key pair", e);
+        }
+    }
+
+    /**
+     * Read input from the terminal as a {@link Period}.
+     * Package protected for testing purposes.
+     */
+    Period readPeriodInput(Terminal terminal, String prompt, Period defaultValue) {
+        defaultValue = tryReadInput(terminal, prompt, defaultValue, input -> {
+            String periodInput = input.replaceAll("[,\\s]", "");
+            if (input.charAt(0) != 'P') {
+                periodInput = "P" + periodInput;
+            }
+            try {
+                return Period.parse(periodInput);
+            } catch (DateTimeParseException e) {
+                terminal.println("Sorry, I do not understand '" + input + "' (" + e.getMessage() + ")");
+                return null;
+            }
+        });
+        return defaultValue;
+    }
+
+    private Integer readKeySize(Terminal terminal, int keySize) {
```

Comment:
Since we go through the process of validating key length, should we explicitly allow only 2048, 3072, 4096 ? If users want to do something extreme or insecure, they can use off-the-os tools for this .

Reply:
Good question. It's a fine balance - we want this to be a guided & easy to use tool that makes it hard to get things wrong, but also let users fit into their existing security policies.
I think you might be right though - or at least another warning/prompt "That's a bad idea, are you really sure"

Reply:
I now enforce that it's a multiple of 1024, in the range [1024, 8192].